man/man1/get.1
.\"	@(#)get.1	6.3 (Berkeley) 10/21/04
.tr ~
.tr $%
.if t .tr #\fB\(em\fP
.if n .tr #-
.if n .ds D " -- 
.if t .ds D _
.tr @|
.nr f 0
.bd S B 3
.de SP
.if n .ul
\%[\fB\(em\\$1\fR\\c
.if n .ul 0
\\$2\\$3
..
.de SF
.if n .ul
\%[\fB\(em\\$1\fR]
.if n .ul 0
..
.de ZZ
.hc ^
\fB^...\fR
.hc
..
.de AR
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\fB\(em\\$1\\fR
\\$2 \\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de C1
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\\$1
\\$2 \\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de A1
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\fB\(em\\$1\fR[\fI\\$2\fR]
\\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de A2
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\fB\(em\\$1\fI\\$2\fR
\\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de FI
.PP
.TP 30
\\$1
\\$2
.i0
..
.ds F)  \fB\s-1FILES\s0\fR
.ds W)  \fB\s-1WARNINGS\s0\fR
.ds X)  \fB\s-1EXAMPLES\s0\fR
.ds T)  \fB\s-1TABLE\s0\fR
.ds K)  \fB\s-1DATA KEYWORDS\s0\fR
.ds D)  \fB\s-1DDDDD\s0\fR
.ds M)  \fB\s-1MR\s0\fR
.ds R)  \fB\s-1RELEASE NUMBER\s0\fR
.ds S)  \fB\s-1SCCS\s0\fR
.ds I)  \fB\s-1SID\s0\fR
.de F1
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 13
.TP 7
\fB\\$1\fI\\$2\fR
\\$3
.nr f 1
..
.de F2
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 13
.TP 7
\fB\\$1[\fI\\$2\fR]
\\$3
.nr f 1
..
.if n .ds )Q '
.if n .ds )G `
.if t .ds )Q \\(aa
.if t .ds )G \\(ga
.if t .ds )S \\|
.TH GET 1 "October 21, 2004"
.SH NAME
get \- get a version of an \s-1SCCS\s+1 file
.SH SYNOPSIS
.na
.B get
.SP r SID ]
.SP c cutoff ]
.SP i list ]
.SP x list ]
.SP a seq~no. ]
.SF k
.SF e
.if n .ul
[#l[p]]
.if n .ul 0
.SF p
.SF m
.SF n
.SF s
.SF b
.SF g
.SF t
file
.ZZ
.ad
.SH DESCRIPTION
.I Get
generates creates an \s-1ASCII\s0 text file from
each named \*(S) file according to the specifications given
by its keyletter arguments,
which begin with \fB\(em\fR.
The arguments
may be specified in any order,
but all keyletter arguments apply to all named \*(S) files.
If a directory is named,
.I get
behaves as though each file in the directory were
specified as a named file,
except that non-\*(S) files
(last component of the pathname does not begin with \fBs.\fR)
and unreadable files
are silently ignored.
If a name of \fB\(em\fR is given, the standard input is read;
each line of the standard input is taken to be the name of an \*(S) file
to be processed.
Again, non-\*(S) files and unreadable files are silently ignored.
.PP
The generated text is normally written into a file called the
.I g-file
whose name is derived from the \*(S) file name by simply
removing the leading \fBs.\fR
(See also \*(F), below).
.PP
Each of the keyletter arguments is explained below
as though only one named \*(S) file is to be processed,
but the effects of any keyletter argument applies independently to
each named file.
.A2 r SID The
.IR \s-1S "CCS " ID\s0 entification
string (\*(I)) of the version (delta)
of an \*(S) file to be retrieved.
\*(T) 1 shows, for the most useful cases, what version
of an \*(S) file is retrieved (as well as the \*(I)
of the version to be eventually created by
.IR delta\^ (1)
if the
.B \(eme
keyletter is also used),
as a function of the \*(I) specified.
.A2 c cutoff \fICutoff\fR
date-time,
in the form
\s-1YY[MM[DD[HH[MM[SS]]]]]\s0.
No changes (deltas) to the \*(S) file which were created after
the specified \fIcutoff\fR date-time are included in the generated
\s-1ASCII\s0 text file.
Units omitted from the date-time default
to their maximum possible values;
that is,
``\fB#c\fR7502''
is equivalent to
``\fB#c\fR750228235959''.
Any number of non-numeric characters may separate
the various 2 digit pieces of the \fIcutoff\fR date-time.
This feature allows one to specify a \fIcutoff\fR date in the form:
``"\fB#c\fR77/2/2 9:22:25"''.
Note that this implies that one may use the
$E$ and $U$ identification keywords
(see below)
for nested
.I gets
within,
say
the input to a
.IR send\^ (1)
command\fB:\fR
.PP
.if t .ti +10
.ie n .ti +20
.tr ~~
~!get  "#c$E$  $U$"  s\*.file
.br
.tr ~
.AR e Indicates
that the
.I get
is for the purpose of editing or making a change
(delta) to the \*(S) file via a subsequent use of
.IR delta\^ (1).
The
.B \(eme
keyletter used in a
.I get
for a particular version (\*(I)) of the \*(S) file prevents
further
.I get\c
s for editing on the same \*(I) until
.I delta
is executed or the
.B j
(joint edit) flag is set in the \*(S) file
(see
.IR admin\^ (1)).
Concurrent use of the
.B \(eme
keyletter in
.I get\c
s
for different \*(I)s is always allowed.
.C1 ~ If
the
.I g-file
generated by
.I get
with an
.B \(eme
keyletter is accidentally ruined in the process of editing it,
it may be regenerated by re-executing the
.I get
command with the
.B \(emk
keyletter
in place of the \fB\(eme\fP keyletter.
.C1 ~ \*(S)
file protection specified via the
ceiling, floor, and authorized user list
stored in the \*(S) file
(see
.IR admin\^ (1))
are enforced when the
.B \(eme
keyletter is used.
.AR b Used
with the
.B \(eme
keyletter to indicate that the new delta should have an \*(I) in a new branch
as shown in \*(T) 1.
This keyletter is allowed only if the
.B b
flag exists in the file
(see
.IR admin\^ (1),
\*(T) 1).
.A2 i list A
\fIlist\fR of deltas to be included
(forced to be applied)
in the creation of the generated file.
The \fIlist\fR has the following syntax:
.F1 ~ ~ <list>
::= <range> \(or <list> , <range>
.br
<range> ::= \*(I) \(or \*(I) - \*(I)
.C1 ~ \*(I),
the \*(S) Identification of a delta, may
be in any form shown in the ``\*(I) Specified'' column of
\*(T) 1.
Partial \*(I)s are interpreted as shown in the ``\*(I) Retrieved''
column of \*(T) 1.
.A2 x list A
\fIlist\fR of deltas to be excluded
(forced not to be applied)
in the creation of the generated file.
See the
.B \(emi
keyletter for the \fIlist\fR format.
.AR k Suppresses
replacement of identification keywords
(see below)
in the retrieved text
by their value.
The
.B \(emk
keyletter
is implied by the
.B \(emi\fR,\fB
.B \(emx
and
.B \(eme
keyletters.
.A1 l \fBp\fR Causes
a delta
summary to be written into an
.I l-file.
If
.B #lp
is used
then an
.I l-file
is not created;
the delta summary is written on the standard output instead.
See \*(F) for the format of the
.I l-file.
.AR p Causes
the text retrieved
from the \*(S) file to
be written on the standard output.
No
.I g-file
is created.
All output which normally goes to the standard output
goes to file descriptor 2 instead,
unless the
.B \(ems
keyletter is used,
in which case it disappears.
.AR s Suppresses
all output normally written on the
standard output.
However,
fatal error messages (which always go to file descriptor
2) remain unaffected.
.AR m Causes
each text line retrieved from the \*(S) file
to be preceded by the
\*(I)
of the delta that inserted the text line in the \*(S) file.
The format is:
\*(I),
followed by a horizontal tab, followed by the text line.
.AR n Causes
each generated text line to be preceded with the
$M$ identification keyword value
(see below).
The format is:
$M$ value,
followed by a horizontal tab,
followed by the text line.
When both the
.B \(emm
and
.B \(emn
keyletters are used,
the format is:
$M$ value,
followed by a horizontal tab,
followed by the
.B \(emm
keyletter generated format.
.AR g Suppresses
the actual retrieval of text from the \*(S) file.
It is primarily used to generate an
.I l-file,
or to verify the existence of a particular \*(I).
.AR t Used
to access the most recently created
(``top'')
delta in a given release
(e\fB.\fR g\fB.,\fR
.B \(emr\c
1)\fB,\fR or
release and level
(e\fB.\fR g\fB.,\fR
.B \(emr\c
1\fB.\fR2)\fB.\fR
.A2 a seq\~no. The
delta sequence number of the \*(S) file delta (version) to be retrieved
(see
.I sccsfile\c
(5)).
This keyletter is used by the
.IR comb\^ (1)
command;
it is not a generally useful keyletter,
and users should not use it.

usr.bin/sccscmds/cmd/get.c
#include <sys/param.h>		/* for MAXPATHLEN */
#undef MAX
# include	"../hdr/defines.h"
# include	"../hdr/had.h"
# include	<sys/dir.h>

static char Sccsid[] = "@(#)get.c	4.11	7/13/2011";

int	Debug = 0;
struct packet gpkt;
struct sid sid;
unsigned	Ser;
int	num_files;
char	had[26];
char	*ilist, *elist, *lfile;
long	cutoff = 0X7FFFFFFFL;	/* max positive long */
int verbosity;
char	Gfile[MAXNAMLEN + 3];
char	Mod[MAXNAMLEN + 3];		/* should be as large as Gfile? */
char	*Type;
int	Did_id;

main(argc,argv)
int argc;
register char *argv[];
{
	register int i;
	register char *p;
	char c;
	int testmore;
	extern int Fcnt;
	extern get();

	Fflags = FTLEXIT | FTLMSG | FTLCLN;
	for(i=1; i<argc; i++)
		if(argv[i][0] == '-' && (c=argv[i][1])) {
			p = &argv[i][2];
			testmore = 0;
			switch (c) {

			case 'a':
				if (!p[0]) {
					argv[i] = 0;
					continue;
				}
				Ser = patoi(p);
				break;
			case 'r':
				if (!p[0]) {
					argv[i] = 0;
					continue;
				}
				chksid(sid_ab(p,&sid),&sid);
				break;
			case 'c':
				if (!p[0]) {
					argv[i] = 0;
					continue;
				}
				if (date_ab(p,&cutoff))
					fatal("bad date/time (cm5)");
				break;
			case 'l':
				lfile = p;
				break;
			case 'i':
				if (!p[0]) {
					argv[i] = 0;
					continue;
				}
				ilist = p;
				break;
			case 'x':
				if (!p[0]) {
					argv[i] = 0;
					continue;
				}
				elist = p;
				break;
			case 'b':
			case 'g':
			case 'e':
			case 'p':
			case 'k':
			case 'm':
			case 'n':
			case 's':
			case 't':
				testmore++;
				break;
			default:
				fatal("unknown key letter (cm1)");
			}

			if (testmore) {
				testmore = 0;
				if (*p) {
					sprintf(Error,"value after %c arg (cm8)",c);
					fatal(Error);
				}
			}
			if (had[c - 'a']++)
				fatal("key letter twice (cm2)");
			argv[i] = 0;
		}
		else num_files++;

	if(num_files == 0)
		fatal("missing file arg (cm3)");
	if (HADE && HADM)
		fatal("e not allowed with m (ge3)");
	if (HADE || HADI || HADX)
		HADK = 1;
	if (!HADS)
		verbosity = -1;
	setsig();
	Fflags &= ~FTLEXIT;
	Fflags |= FTLJMP;
	for (i=1; i<argc; i++)
		if (p=argv[i])
			do_file(p,get);
	exit(Fcnt ? 1 : 0);
}


get(file)
{
	register char *p;
	register unsigned ser;
	extern char had_dir, had_standinp;
	extern char *Sflags[];
	struct stats stats;
	char	str[32];

	if (setjmp(Fjmp))
		return;
	sinit(&gpkt,file,1);
	gpkt.p_ixuser = (HADI | HADX);
	gpkt.p_reqsid.s_rel = sid.s_rel;
	gpkt.p_reqsid.s_lev = sid.s_lev;
	gpkt.p_reqsid.s_br = sid.s_br;
	gpkt.p_reqsid.s_seq = sid.s_seq;
	gpkt.p_verbose = verbosity;
	gpkt.p_stdout = (HADP ? stderr : stdout);
	gpkt.p_cutoff = cutoff;
	gpkt.p_lfile = lfile;
	copy(auxf(gpkt.p_file,'g'),Gfile);

	if (gpkt.p_verbose && (num_files > 1 || had_dir || had_standinp))
		fprintf(gpkt.p_stdout,"\n%s:\n",gpkt.p_file);
	if (dodelt(&gpkt,&stats,0,0) == 0)
		fmterr(&gpkt);
	finduser(&gpkt);
	doflags(&gpkt);
	if (!HADA)
		ser = getser(&gpkt);
	else {
		if ((ser = Ser) > maxser(&gpkt))
			fatal("serial number too large (ge19)");
		bcopy(&gpkt.p_idel[ser].i_sid, &gpkt.p_gotsid, sizeof(sid));
		if (HADR && sid.s_rel != gpkt.p_gotsid.s_rel) {
			bzero(&gpkt.p_reqsid, sizeof(gpkt.p_reqsid));
			gpkt.p_reqsid.s_rel = sid.s_rel;
		}
		else
			bcopy(&gpkt.p_gotsid, &gpkt.p_reqsid, sizeof(sid));
	}
	doie(&gpkt,ilist,elist,0);
	setup(&gpkt,ser);
	if (!(Type = Sflags[TYPEFLAG - 'a']))
		Type = Null;
	if (!(HADP || HADG) && writable(Gfile)) {
		sprintf(Error,"writable `%s' exists (ge4)",Gfile);
		fatal(Error);
	}
	if (gpkt.p_verbose) {
		sid_ba(&gpkt.p_gotsid,str);
		fprintf(gpkt.p_stdout,"%s\n",str);
	}
	if (HADE) {
		if (!HADR)
			bcopy(&gpkt.p_gotsid,&gpkt.p_reqsid,
			     sizeof(gpkt.p_reqsid));
		newsid(&gpkt,Sflags[BRCHFLAG - 'a'] && HADB);
		permiss(&gpkt);
		wrtpfile(&gpkt,ilist,elist);
	}
	setuid(getuid());
	if (HADL)
		gen_lfile(&gpkt);
	if (HADG) {
		fclose(gpkt.p_iop);
		xfreeall();
		return;
	}
	flushto(&gpkt,EUSERTXT,1);
	idsetup(&gpkt);
	gpkt.p_chkeof = 1;
	Did_id = 0;
	while(readmod(&gpkt)) {
		if (gpkt.p_gout == 0) {
			if (HADP)
				gpkt.p_gout = stdout;
			else
				gpkt.p_gout = xfcreat(Gfile,HADK ? 0666 : 0444);
		}
		if (HADN)
			fprintf(gpkt.p_gout,"%s\t",Mod);
		if (HADM) {
			sid_ba(&gpkt.p_inssid,str);
			fprintf(gpkt.p_gout,"%s\t",str);
		}
		p = gpkt.p_line;
		if (!HADK && any('%',p))
			p = idsubst(&gpkt,p);
		fputs(p,gpkt.p_gout);
	}
	fflush(gpkt.p_gout);
	if (gpkt.p_gout && gpkt.p_gout != stdout)
		fclose(gpkt.p_gout);
	if (gpkt.p_verbose)
		fprintf(gpkt.p_stdout,"%u lines\n",gpkt.p_glnno);
	if (!Did_id && !HADK)
		if (Sflags[IDFLAG - 'a'])
			fatal("no id keywords (cm6)");
		else if (gpkt.p_verbose)
			fprintf(stderr,"No id keywords (cm7)\n");
	xfreeall();
}

writable(fn)
char *fn;
{
	struct stat s;

	return (stat(fn, &s) >= 0 && (s.st_mode & 0222) != 0);
}


newsid(pkt,branch)
register struct packet *pkt;
int branch;
{
	int chkbr;

	chkbr = 0;
	if (pkt->p_reqsid.s_br == 0) {
		pkt->p_reqsid.s_lev += 1;
		if (sidtoser(&pkt->p_reqsid,pkt) ||
			pkt->p_maxr > pkt->p_reqsid.s_rel || branch) {
				pkt->p_reqsid.s_rel = pkt->p_gotsid.s_rel;
				pkt->p_reqsid.s_lev = pkt->p_gotsid.s_lev;
				pkt->p_reqsid.s_br = pkt->p_gotsid.s_br + 1;
				pkt->p_reqsid.s_seq = 1;
				chkbr++;
		}
	}
	else if (pkt->p_reqsid.s_seq == 0 && !branch)
		pkt->p_reqsid.s_seq = pkt->p_gotsid.s_seq + 1;
	else {
		pkt->p_reqsid.s_seq += 1;
		if (branch || sidtoser(&pkt->p_reqsid,pkt)) {
			pkt->p_reqsid.s_br += 1;
			pkt->p_reqsid.s_seq = 1;
			chkbr++;
		}
	}
	if (chkbr)
		while (sidtoser(&pkt->p_reqsid,pkt))
			pkt->p_reqsid.s_br += 1;
	if (sidtoser(&pkt->p_reqsid,pkt))
		fatal("internal error in newsid()");
}


enter(pkt,ch,n,sidp)
struct packet *pkt;
char ch;
int n;
struct sid *sidp;
{
	char str[32];
	register struct apply *ap;

	sid_ba(sidp,str);
	if (pkt->p_verbose)
		fprintf(pkt->p_stdout,"%s\n",str);
	ap = &pkt->p_apply[n];
	switch(ap->a_code) {

	case EMPTY:
		if (ch == INCLUDE)
			condset(ap,APPLY,INCLUSER);
		else
			condset(ap,NOAPPLY,EXCLUSER);
		break;
	case APPLY:
		sid_ba(sidp,str);
		sprintf(Error,"%s already included (ge9)",str);
		fatal(Error);
		break;
	case NOAPPLY:
		sid_ba(sidp,str);
		sprintf(Error,"%s already excluded (ge10)",str);
		fatal(Error);
		break;
	default:
		fatal("internal error in get/enter() (ge11)");
		break;
	}
}


gen_lfile(pkt)
register struct packet *pkt;
{
	int n;
	int reason;
	char str[32];
	char line[BUFSIZ];
	struct deltab dt;
	FILE *in;
	FILE *out;

	in = xfopen(pkt->p_file,0);
	if (*pkt->p_lfile)
		out = stdout;
	else
		out = xfcreat(auxf(pkt->p_file,'l'),0444);
	fgets(line,sizeof(line),in);
	while (fgets(line,sizeof(line),in) != NULL && line[0] == CTLCHAR && line[1] == STATS) {
		fgets(line,sizeof(line),in);
		del_ab(line,&dt);
		if (dt.d_type == 'D') {
			reason = pkt->p_apply[dt.d_serial].a_reason;
			if (pkt->p_apply[dt.d_serial].a_code == APPLY) {
				putc(' ',out);
				putc(' ',out);
			}
			else {
				putc('*',out);
				if (reason & IGNR)
					putc(' ',out);
				else
					putc('*',out);
			}
			switch (reason & (INCL | EXCL | CUTOFF)) {
	
			case INCL:
				putc('I',out);
				break;
			case EXCL:
				putc('X',out);
				break;
			case CUTOFF:
				putc('C',out);
				break;
			default:
				putc(' ',out);
				break;
			}
			putc(' ',out);
			sid_ba(&dt.d_sid,str);
			fprintf(out,"%s\t",str);
			date_ba(&dt.d_datetime,str);
			fprintf(out,"%s %s\n",str,dt.d_pgmr);
		}
		while ((n = fgets(line,sizeof(line),in)) != NULL)
			if (line[0] != CTLCHAR)
				break;
			else {
				switch (line[1]) {

				case EDELTAB:
					break;
				default:
					continue;
				case MRNUM:
				case COMMENTS:
					if (dt.d_type == 'D')
						fprintf(out,"\t%s",&line[3]);
					continue;
				}
				break;
			}
		if (n == NULL || line[0] != CTLCHAR)
			break;
		putc('\n',out);
	}
	fclose(in);
	if (out != stdout)
		fclose(out);
}

char	Curdate[20];
char	*Curtime;
char	Gdate[11];
char	Chgdate[20];
char	*Chgtime;
char	Gchgdate[11];
char	Qchgdate[30];
char	Sid[32];
char	Olddir[MAXPATHLEN+1];
char	Pname[MAXPATHLEN+1];
char	Dir[MAXPATHLEN+1];

idsetup(pkt)
register struct packet *pkt;
{
	extern long Timenow;
	register int n;
	register char *p;

	date_ba(&Timenow,Curdate);
	Curtime = &Curdate[11];
	Curdate[10] = 0;
	copy(pkt->p_file,Dir);
	dname(Dir);
	if(getwd(Olddir) == 0)
		fatal("getwd failed (ge20)");
	if(chdir(Dir) != 0)
		fatal("cannot change directory (ge22)");
	if(getwd(Pname) == 0)
		fatal("getwd failed (ge21)");
	if(chdir(Olddir) != 0)
		fatal("cannot change directory (ge23)");
	makgdate(Curdate,Gdate);
	for (n = maxser(pkt); n; n--)
		if (pkt->p_apply[n].a_code == APPLY)
			break;
	if (n)
		date_ba(&pkt->p_idel[n].i_datetime,Chgdate);
	Chgtime = &Chgdate[11];
	Chgdate[10] = 0;
	makgdate(Chgdate,Gchgdate);
	makqdate(Gchgdate,Qchgdate);
	sid_ba(&pkt->p_gotsid,Sid);
	if (p = Sflags[MODFLAG - 'a'])
		copy(p,Mod);
	else
		copy(Gfile,Mod);
}


makgdate(old,new)
register char *old, *new;
{
	if ((*new = old[5]) != '0')
		new++;
	*new++ = old[6];
	*new++ = '/';
	if ((*new = old[8]) != '0')
		new++;
	*new++ = old[9];
	*new++ = '/';
	*new++ = old[0];
	*new++ = old[1];
	*new++ = old[2];
	*new++ = old[3];
	*new = 0;
}

makqdate(old,new)
register char *old, *new;
{
	static	char *months[12] =
	    { "January", "February", "March", "April", "May", "June", "July",
	      "August", "September", "October", "November", "December" };

	strcpy(new, months[atoi(old)-1]);
	while (*new != '\0')
		new++;
	while (*old++ != '/')
		;
	*new++ = ' ';
	*new++ = *old++;
	if (*old != '/')
		*new++ = *old++;
	old++;
	*new++ = ',';
	*new++ = ' ';
	*new++ = *old++;
	*new++ = *old++;
	*new++ = *old++;
	*new++ = *old++;
	*new = '\0';
}

static char Zkeywd[5] = "@(#)";


idsubst(pkt,line)
register struct packet *pkt;
char line[];
{
	static char tline[BUFSIZ];
	static char str[32];
	register char *lp, *tp;
	extern char *Type;
	extern char *Sflags[];

	tp = tline;
	for(lp=line; *lp != 0; lp++) {
		if(lp[0] == '%' && lp[1] != 0 && lp[2] == '%') {
			switch(*++lp) {

			case 'M':
				tp = trans(tp,Mod);
				break;
			case 'R':
				sprintf(str,"%u",pkt->p_gotsid.s_rel);
				tp = trans(tp,str);
				break;
			case 'L':
				sprintf(str,"%u",pkt->p_gotsid.s_lev);
				tp = trans(tp,str);
				break;
			case 'B':
				sprintf(str,"%u",pkt->p_gotsid.s_br);
				tp = trans(tp,str);
				break;
			case 'S':
				sprintf(str,"%u",pkt->p_gotsid.s_seq);
				tp = trans(tp,str);
				break;
			case 'D':
				tp = trans(tp,Curdate);
				break;
			case 'H':
				tp = trans(tp,Gdate);
				break;
			case 'T':
				tp = trans(tp,Curtime);
				break;
			case 'E':
				tp = trans(tp,Chgdate);
				break;
			case 'G':
				tp = trans(tp,Gchgdate);
				break;
			case 'Q':
				tp = trans(tp,Qchgdate);
				break;
			case 'U':
				tp = trans(tp,Chgtime);
				break;
			case 'Z':
				tp = trans(tp,Zkeywd);
				break;
			case 'Y':
				tp = trans(tp,Type);
				break;
			case 'W':
				tp = trans(tp,Zkeywd);
				tp = trans(tp,Mod);
				*tp++ = '\t';
			case 'I':
				tp = trans(tp,Sid);
				break;
			case 'P':
				tp = trans(tp,Pname);
				*tp++ = '/';
				tp = trans(tp,(sname(pkt->p_file)));
				break;
			case 'F':
				tp = trans(tp,pkt->p_file);
				break;
			case 'C':
				sprintf(str,"%u",pkt->p_glnno);
				tp = trans(tp,str);
				break;
			case 'A':
				tp = trans(tp,Zkeywd);
				tp = trans(tp,Type);
				*tp++ = ' ';
				tp = trans(tp,Mod);
				*tp++ = ' ';
				tp = trans(tp,Sid);
				tp = trans(tp,Zkeywd);
				break;
			default:
				*tp++ = '%';
				*tp++ = *lp;
				continue;
			}
			lp++;
		}
		else
			*tp++ = *lp;
	}

	*tp = 0;
	return(tline);
}


trans(tp,str)
register char *tp, *str;
{
	Did_id = 1;
	while(*tp++ = *str++)
		;
	return(tp-1);
}


clean_up(n)
{
	if (gpkt.p_file[0])
		unlockit(auxf(gpkt.p_file,'z'),getpid());
	if (gpkt.p_iop)
		fclose(gpkt.p_iop);
	xfreeall();
}


wrtpfile(pkt,inc,exc)
register struct packet *pkt;
char *inc, *exc;
{
	char line[64], str1[32], str2[32];
	char *user;
	FILE *in, *out;
	struct pfile pf;
	register char *p;
	int fd;
	int i;
	extern long Timenow;

	user = logname();
	if (lockit(auxf(pkt->p_file,'z'),2,getpid()))
		fatal("cannot create lock file (cm4)");
	if (exists(p = auxf(pkt->p_file,'p'))) {
		fd = xopen(p,2);
		in = fdopen(fd,"r");
		while (fgets(line,sizeof(line),in) != NULL) {
			p = line;
			p[length(p) - 1] = 0;
			pf_ab(p,&pf,0);
			if ((pf.pf_gsid.s_rel == pkt->p_gotsid.s_rel &&
				pf.pf_gsid.s_lev == pkt->p_gotsid.s_lev &&
				pf.pf_gsid.s_br == pkt->p_gotsid.s_br &&
				pf.pf_gsid.s_seq == pkt->p_gotsid.s_seq) ||
				(pf.pf_nsid.s_rel == pkt->p_reqsid.s_rel &&
				pf.pf_nsid.s_lev == pkt->p_reqsid.s_lev &&
				pf.pf_nsid.s_br == pkt->p_reqsid.s_br &&
				pf.pf_nsid.s_seq == pkt->p_reqsid.s_seq)) {
				fclose(in);
				sprintf(Error,"being edited: `%s' (ge17)",line);
				fatal(Error);
				}
			if (!equal(pf.pf_user,user))
				fprintf(stderr,"WARNING: being edited: `%s' (ge18)\n",line);
		}
		out = fdopen(dup(fd),"w");
		fclose(in);
	}
	else
		out = xfcreat(p,0666);
	fseek(out,0L,2);
	sid_ba(&pkt->p_gotsid,str1);
	sid_ba(&pkt->p_reqsid,str2);
	date_ba(&Timenow,line);
	fprintf(out,"%s %s %s %s",str1,str2,user,line);
	if (inc)
		fprintf(out," -i%s",inc);
	if (exc)
		fprintf(out," -x%s",exc);
	fprintf(out,"\n");
	fclose(out);
	if (pkt->p_verbose)
		fprintf(pkt->p_stdout,"new delta %s\n",str2);
	unlockit(auxf(pkt->p_file,'z'),getpid());
}


getser(pkt)
register struct packet *pkt;
{
	register struct idel *rdp;
	int n, ser, def;
	char *p;
	extern char *Sflags[];

	def = 0;
	if (pkt->p_reqsid.s_rel == 0) {
		if (p = Sflags[DEFTFLAG - 'a'])
			chksid(sid_ab(p, &pkt->p_reqsid), &pkt->p_reqsid);
		else {
			pkt->p_reqsid.s_rel = MAX;
			def = 1;
		}
	}
	ser = 0;
	if (pkt->p_reqsid.s_lev == 0) {
		for (n = maxser(pkt); n; n--) {
			rdp = &pkt->p_idel[n];
			if ((rdp->i_sid.s_br == 0 || HADT) &&
				pkt->p_reqsid.s_rel >= rdp->i_sid.s_rel &&
				rdp->i_sid.s_rel > pkt->p_gotsid.s_rel) {
					ser = n;
					pkt->p_gotsid.s_rel = rdp->i_sid.s_rel;
			}
		}
	}
	else if (pkt->p_reqsid.s_br && pkt->p_reqsid.s_seq == 0) {
		for (n = maxser(pkt); n; n--) {
			rdp = &pkt->p_idel[n];
			if (rdp->i_sid.s_rel == pkt->p_reqsid.s_rel &&
				rdp->i_sid.s_lev == pkt->p_reqsid.s_lev &&
				rdp->i_sid.s_br == pkt->p_reqsid.s_br)
					break;
		}
		ser = n;
	}
	else {
		ser = sidtoser(&pkt->p_reqsid,pkt);
	}
	if (ser == 0)
		fatal("nonexistent sid (ge5)");
	rdp = &pkt->p_idel[ser];
	bcopy(&rdp->i_sid, &pkt->p_gotsid, sizeof(pkt->p_gotsid));
	if (def || (pkt->p_reqsid.s_lev == 0 && pkt->p_reqsid.s_rel == pkt->p_gotsid.s_rel))
		bcopy(&pkt->p_gotsid, &pkt->p_reqsid, sizeof(pkt->p_gotsid));
	return(ser);
}


/* Null routine to satisfy external reference from dodelt() */

escdodelt()
{
}
